### DISCUSSION TITLE
add builtin weakmap version of atomFamily?

### DISCUSSION THREAD

**Discussion by himself65** [pmndrs/jotai#2239] [General] [answered]
In my source code, there are so many weak maps from an object to an atom. I'm not sure if the jotai could add it as the builtin utility.

**Comment by dai-shi** [comment, marked as answer]
I have been interested in the idea of `weakAtomFamily`. It can be the Jotai way.
Would it be pretty simple?

```js
export function weakAtomFamily(initializeAtom) {
  const atoms = new WeakMap();
  return (obj) => {
    if (!atoms.has(obj)) {
      atoms.set(obj, initializeAtom(obj));
    }
    return atoms.get(obj);
  };
}
```

Maybe, it's better in a recipe.
(tbh, I'd like to move atomFamily to recipes in docs, if possible.)

**Reply by probeiuscorp** [reply]
I've been using this pattern extensively in the last few months and I think it should be recommended in the docs. In particular, I think it 'solves' the DX gap between global atoms and atoms as props.

One of the best parts of Jotai is writing all your state at the global level:
```javascript
const listAtom = atom([1, 2, 3, 4])
const lengthAtom = atom((get) => get(listAtom).length)
function SomeComponent() {
  const [length] = useAtom(lengthAtom)
}
```
But when `listAtom` is received from props, your code has to change dramatically:
```javascript
function SomeComponent({ listAtom }) {
  const lengthAtom = useMemo((listAtom) => atom((get) => get(listAtom).length), [listAtom])
  const [length] = useAtom(lengthAtom)
}
```
However, with `weakAtomFamily`:
```javascript
const getLengthAtom = weakAtomFamily((listAtom) => atom((get) => get(listAtom).length))
function SomeComponent({ listAtom }) {
  const lengthAtom = getLengthAtom(listAtom)
  const [length] = useAtom(listAtom)
}
```
As an added bonus, `getLengthAtom`:
- doesn't depend on the semantic guarantee (/lack thereof) of `useMemo` (which might not affect this readonly atom, but I've used many read/write atoms with this pattern)
- can be called conditionally, and
- can be accessed from any code path where `listAtom` can be (so shareable between sibling components!)

I would prefer a different name though. I can't think of anything better than `stable()` but I believe this pattern deserves a short name. It also isn't particular to atoms or Jotai so I think it'd be nicest not to include atom in the name.

**Reply by dai-shi** [reply]
That's an interesting pattern.
I think I've used atoms-in-atom pattern in such use cases, 
but atom-as-weak-map-key looks more intuitive.

Can you create a small working example with https://csb.jotai.org ? We can then compare both patterns.

**Reply by probeiuscorp** [reply]
I've made a [working example](https://codesandbox.io/p/sandbox/jotai-demo-stable-58xzkv?file=%2Fsrc%2FREADME%3A17%2C48) for some of the use cases of `stable()` I have in mind. For the atom-in-atom use case, LayerName would be the place to look.

**Reply by dai-shi** [reply]
Thanks for the example. It's also atom-in-atom in a sense. Some notes:
- `stable()` is a general util, not tied to jotai, which is good.
- this would be a good pattern in recipes or guides in jotai docs.
- i would create two atoms at the same time and put the pair in the layer object, but i understand your approach is somewhat better DX.
- we have `jotai-scope` which is for slightly different problem, but this might be a good insight for it.

**Reply by grzesiek-ds** [reply]
@dai-shi are You planning to work on `weakAtomFamily`? If not then is this topic still open and if yes, then re You accepting any contributions in this field, or You would prefer to keep it outside of `jotai`?

**Reply by dai-shi** [reply]
I am interested in `weakAtomFamily` (more than `atomFamily`), but I'm not planning to work on it (because I don't really understand the usage requirement.) We don't plan to add it in `jotai/utils` (I actually would like to move `atomFamily` out too, eventually). For the moment, it should be in the recipe in docs and/or 3rd-party libs.

**Reply by jaens** [reply]
Here's a use case description:

`atomFamily` requires "stable" keys (eg. database IDs from the backend) so it would not leak memory.

Sometimes though, it's convenient to calculate *and* cache atoms based on a parameter that is an actual object (which can eg. contain extra information), but the object could be "temporary"/"dynamic" (eg. created purely based on client-side data, or re-created multiple times from backend data on periodic refresh).

In this case, instead of trying to invent a stable ID (which is possible in 90% of cases, but can be inconvenient), temporary objects could be directly used as a family keys, but the `Map` would have to `Weak` to prevent memory leaks.

**Reply by dai-shi** [reply]
Does the implementation in https://github.com/pmndrs/jotai/discussions/2239#discussioncomment-7494465 work for your case?

**Reply by jaens** [reply]
<details>
<summary>
Yes, I use the slightly optimized:
</summary>

```typescript
export function atomFamilyWeak<K extends WeakKey, V>(create: (key: K) => V): (key: K) => V {
    const cache = new WeakMap<K, V>();
    return (key) => {
        let value;
        if (!cache.has(key)) {
            value = create(key);
            cache.set(key, value);
        } else {
            value = cache.get(key)!;
        }
        return value;
    };
}
```
</details>
